### PR TITLE
Allow to find classes in multiple directories & improvements

### PR DESCRIPTION
**Pull request by rustamwin**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️/❌
| Fixed issues  |

### CODE REVIEW

File: src/Classifier.php [hunk 1]
```diff
@@ -105,7 +117,7 @@ public function find(): iterable
                 }
             }
 
-            if (($this->parent !== null) && !is_subclass_of($className, $this->parent)) {
+            if (($this->targetClass !== null) && !is_subclass_of($className, $this->targetClass)) {
```
Comment (xepozz) [inline src/Classifier.php:120]: `targetClass` looks like a class that you exactly want to find, but actually it finds also inheriting classes

File: src/Classifier.php [hunk 2]
```diff
@@ -19,12 +19,17 @@ final class Classifier
...
+    private ?string $targetClass = null;
+    /**
+     * @var string[]
+     */
+    private array $dirs;
 
-    public function __construct(private string $directory)
+    public function __construct(string ...$directory)
     {
+        $this->dirs = $directory;
```
Comment (xepozz) [inline src/Classifier.php:32]:
```suggestion
        $this->directories = $directory;
```

**Review by xepozz — CHANGES_REQUESTED** [review]

File: src/Classifier.php [hunk 3]
```diff
@@ -79,6 +79,10 @@ public function find(): iterable
         $this->scanFiles();
 
         $classesToFind = get_declared_classes();
+        $directories = array_map(
+            static fn($directory) => DIRECTORY_SEPARATOR === '\\' ? str_replace('/', '\\', $directory) : $directory,
```
Comment (xepozz) [inline src/Classifier.php:83]: Please check `DIRECTORY_SEPARATOR === '\\'` outside the function and use `str_replace` with array arguments:
```php

$directories = $this->directories;

$isWindows = DIRECTORY_SEPARATOR === '\\';
if ($isWindows) {
    $directories = str_replace('/', '\\', $this->directories);
}
```

**Review by xepozz — CHANGES_REQUESTED** [review]

File: src/Classifier.php (hunk 3 above)
Comment (xepozz) [reply src/Classifier.php]: @rustamwin `str_replace` support batch replacing, no need to iterate and replace individually

File: src/Classifier.php [hunk 4]
```diff
@@ -19,12 +19,17 @@ final class Classifier
      */
     private array $attributes = [];
     /**
-     * @psalm-var class-string
+     * @var class-string
```
Comment (vjik) [inline src/Classifier.php:22]: `class-string` is psalm-specific type

File: src/Classifier.php [hunk 5]
```diff
@@ -59,28 +64,42 @@ public function withAttribute(string ...$attributes): self
         return $new;
     }
 
+    /**
+     * @return iterable<class-string>
```
Comment (vjik) [inline src/Classifier.php:68]:
```suggestion
     * @psalm-return iterable<class-string>
```

File: src/Classifier.php [hunk 6]
```diff
@@ -59,28 +64,42 @@ public function withAttribute(string ...$attributes): self
...
+            /** @var string[] $directories */
+            $directories = str_replace('/', '\\', $directories);
         }
 
         foreach ($classesToFind as $className) {
             $reflection = new ReflectionClass($className);
-            $filePath = $reflection->getFileName();
-            if ($filePath === false || !str_starts_with($filePath, $baseDirectory)) {
+
+            if (!$reflection->isUserDefined()) {
```
Comment (vjik) [inline src/Classifier.php:93]: Directory is required for search classes? If yes need modify constructor and make one parameter required.

**Review by vjik — CHANGES_REQUESTED** [review]

**Review by xepozz — APPROVED** [review]

**Review by vjik — APPROVED** [review]

**Pull request merged**